src/lib.rs

```diff
@@ -315,6 +316,191 @@ impl<T: Hash + Clone + Debug + Sync + Send> Mphf<T> {
 }
 
 
+////////////////////////////////
+// Adding Support for new BoomHashMap object
+////////////////////////////////
+// TODO: Don't like copy pasting three versions of Boom, has to be a better way
+pub struct BoomHashMap<K: Clone + Hash + Debug, D> {
+    mphf: Mphf<K>,
+    keys: Vec<K>,
+    values: Vec<D>
+}
+
+impl<K, D> BoomHashMap<K, D>
+where K: Clone + Hash + Debug + PartialEq, D: Debug {
+    pub fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Kmer {{ id: {:?}, Data: {:?} }} \n Lengths {}, {}",
+            self.keys,
+            self.values,
+            self.keys.len(),
+            self.values.len()
+        )
+    }
+
+    pub fn new(mut keys_: Vec<K>, mut data_: Vec<D> ) -> BoomHashMap<K, D> {
+        let mphf_ = Mphf::new(1.7, &keys_, None);
+        // trick taken from :
+        // https://github.com/10XDev/cellranger/blob/master/lib/rust/detect_chemistry/src/index.rs#L123
+        for i in 0 .. keys_.len() {
+            loop {
+                let kmer_slot = mphf_.hash(&keys_[i]) as usize;
+                if i == kmer_slot { break; }
+                keys_.swap(i, kmer_slot);
+                data_.swap(i, kmer_slot);
+            }
+        }
+        BoomHashMap{
+            mphf: mphf_,
+            keys: keys_,
+            values: data_,
+        }
+    }
+
+    pub fn get_data_for_kmer(&self, kmer: &K) -> Option<&D> {
+
+        let maybe_pos = self.mphf.try_hash(&kmer);
+        match maybe_pos {
+            Some(pos) => {
+                let hashed_kmer = &self.keys[pos as usize];
+                if *kmer == hashed_kmer.clone() {
+                    Some(&self.values[pos as usize])
+                }
+                else {
+                    None
+                }
+            },
+            None => None,
+        }
+    }
+    pub fn get_kmer_id(&self, kmer: &K) -> Option<usize> {
+
+        let maybe_pos = self.mphf.try_hash(&kmer);
+        match maybe_pos {
+            Some(pos) => {
+                let hashed_kmer = &self.keys[pos as usize];
+                if *kmer == hashed_kmer.clone() {
+                    Some(pos as usize)
+                }
+                else {
+                    None
+                }
+            },
+            None => None,
+        }
+    }
+
+    pub fn num_elem(&self) -> usize {
+        self.keys.len()
+    }
+
+    pub fn get_key(&self, id:usize) -> Option<K> {
+        let max_key_id = self.num_elem();
+        if id > max_key_id {
+            None
+        }
+        else{
+            Some(self.keys[id].clone())
+        }
+    }
+
```

Comment:
At some point we should also add some iterators e.g. iter() and iter_mut() to be similar to HashMap